src/Symfony/Component/DependencyInjection/ContainerBuilder.php

```diff
@@ -418,11 +414,17 @@ public function get($id, $invalidBehavior = ContainerInterface::EXCEPTION_ON_INV
         try {
             $definition = $this->getDefinition($id);
         } catch (ServiceNotFoundException $e) {
-            if (ContainerInterface::EXCEPTION_ON_INVALID_REFERENCE !== $invalidBehavior) {
-                return;
+            if (ContainerInterface::EXCEPTION_ON_INVALID_REFERENCE === $invalidBehavior) {
+                throw $e;
             }
+            $definition = null;
+        }
 
-            throw $e;
+        if (!$this->compiled) {
```

Comment:
Shouldn't this stay before the code trying to get a definition ? It would avoid having to do weird things where you reset the definition variable to return later, which is much less readable.
A synthetic service will not be instantiated from a Definition object anyway.